lisa/sut_orchestrator/hyperv/platform_.py

```diff
@@ -0,0 +1,295 @@
+        )
+        node_capabilities.disk = schema.DiskOptionSettings(
+            data_disk_count=search_space.IntRange(min=0),
+            data_disk_size=search_space.IntRange(min=1),
+        )
+        node_capabilities.network_interface = schema.NetworkInterfaceOptionSettings()
+        node_capabilities.network_interface.max_nic_count = 1
+        node_capabilities.network_interface.nic_count = 1
+        node_capabilities.gpu_count = 0
+        node_capabilities.features = search_space.SetSpace[schema.FeatureSettings](
+            is_allow_set=True,
+            items=[
+                schema.FeatureSettings.create(SerialConsole.name()),
+            ],
+        )
+
+        return node_capabilities
+
+    def _deploy_environment(self, environment: Environment, log: Logger) -> None:
+        self._deploy_nodes(environment, log)
+
+    def _deploy_nodes(self, environment: Environment, log: Logger) -> None:
+        if environment.runbook.nodes_requirement is None:
+            return  # nothing to deploy?
+
+        test_suffix = "".join(random.choice(string.ascii_uppercase) for _ in range(5))
+        vm_name_prefix = f"lisa-{test_suffix}"
+
+        hv = self.server_node.tools[HyperV]
+        default_switch = hv.get_first_switch()
+
+        extra_args = {
+            x.command.lower(): x.args for x in self._hyperv_runbook.extra_args
+        }
+
+        self.console_logger = SerialConsoleLogger(self.server_node)
+
+        for i, node_space in enumerate(environment.runbook.nodes_requirement):
+            node_runbook = node_space.get_extended_runbook(
+                HypervNodeSchema, type(self).type_name()
+            )
+
+            vm_name = f"{vm_name_prefix}-{i}"
+
+            node = environment.create_node_from_requirement(node_space)
+            assert isinstance(node, RemoteNode)
+
+            node.name = vm_name
+
+            node_context = get_node_context(node)
+            node_context.vm_name = vm_name
+            node_context.server_node = self.server_node
+            node_context.vhd_local_path = PurePosixPath(node_runbook.vhd)
+            node_context.vhd_remote_path = PureWindowsPath(
+                self.server_node.working_path / f"{vm_name}-vhd.vhdx"
+            )
+            node_context.console_log_path = PureWindowsPath(
+                self.server_node.working_path / f"{vm_name}-console.log"
+            )
+
+            remote_path = node_context.vhd_remote_path
+            is_zipped = False
+            if node_context.vhd_local_path.suffix == ".zip":
+                remote_path = PureWindowsPath(
+                    self.server_node.working_path / f"{vm_name}-vhd.zip"
+                )
+                is_zipped = True
+
+            log.debug("Copying VHD to server")
+            self.server_node.shell.copy(node_context.vhd_local_path, remote_path)
+            log.debug("Finished copying VHD to server")
+
+            if is_zipped:
```

Comment:
When will it be zipped? It should be handled in transformer, and unzip one time.

Reply:
The idea is to allow specifying a zipped VHD in the runbook. This speeds up the copy of the VHD to the host node. We will copy in zipped format and afterwards unzip it on the host node. `is_zipped` will be true if we detect a .zip extension in the VHD path provided in the runbook.

Reply:
Tested with our ~2GB VHD that, when zipped, comes down to ~300MB. So, copying over the network is much faster.

Reply:
I agree it should be one time. I will make the changes.

Not sure about transformer though. If this is a transformer its output will be a variable that gives the _remote_ VHD path. Then the platform schema needs to be modified to be able to accept a remote VHD path.

Reply:
I made it one time.

Waiting to hear further thoughts from you on the transformer. It will require runbook schema change.

Reply:
There is no such a transformer, but you can create one to extract zip file, and return a path to vhd variable. It doesn't need to change platform schema, because the changes happen on variables.

Reply:
The unzip should happen _after_ copying to the server node. Which means the transformer will have to update the variable with the _remote path_ of the extracted VHD. So, the VHD field in runbook can be interpreted in two ways: it can be a _remote path_ if unzip transformer is used, else it will be _local path_. So there has to be some change in the schema to indicate whether the vhd path is remote or local.

Reply:
It doesn't need to be a remote path. In lisa, there are local node and remote node. When you operate a node, it's always local path. There is an uploader transformer already, but it needs to figure out how to work with host nodes, instead of vm nodes. Otherwise, it needs an unzip transformer. Another way is refer to BareMetal solution, use a build part, and copy images from a specified path, and execute other logic.

Reply:
It needs to be a remote path because that is the whole point of doing this. We want to copy the zipped file over the network and unzip it on the remote node. This way we copy a smaller size over the network thus saving time.

This is the sequence of operations we need:
1. Copy the zipped VHD to the *server node*. Let's say it is copied to `C:\Users\Administrator\lisatest\image.zip`. This path is on the server node i.e. it is a remote path.
2. Unzip the VHD _on the server node_. Let's say it got unzipped at `C:\Users\Administrator\lisatest\image.vhd`.
3. Now use `image.vhd` as the source VHD for all VMs. To do this, the platform needs to be aware of the path to `image.vhd` i.e. `C:\Users\Administrator\lisatest\image.vhd`.

As you can see in this example, the final VHD path is a remote path. If a transformer performs the above operations, it needs to modify the `$(vhd)` variable to have the *remote path* so that the platform is aware of where to find the source VHD.

Reply:
Introduced VHDSource in the latest iteration.